[CONTP-1760] Document configuring Autodiscovery with the DatadogInstrumentation CRD (#37532)
DataDog/documentation: content/en/containers/guide/configure-autodiscovery-with-the-datadoginstrumentation-crd.md

---
title: Configure Autodiscovery with DatadogInstrumentation CRD
description: Configure Autodiscovery checks and logs for Kubernetes workloads through the DatadogInstrumentation custom resource instead of pod annotations.
further_reading:
- link: "/containers/kubernetes/integrations/"
  tag: "Documentation"
  text: "Configure integrations with Autodiscovery"
- link: "/getting_started/containers/autodiscovery/"
  tag: "Documentation"
  text: "Getting Started with Autodiscovery"
- link: "/containers/guide/autodiscovery-examples/"
  tag: "Documentation"
  text: "Autodiscovery scenarios and examples"
- link: "/containers/cluster_agent/"
  tag: "Documentation"
  text: "Datadog Cluster Agent"
---

## Overview

> The `DatadogInstrumentation` custom resource (CR) lets you configure [Autodiscovery][1] checks and logs with a single Kubernetes resource, instead of [pod annotations][2]. With this approach, you can enable, update, and remove integration configurations without editing your Agent or application and triggering a rollout.

Reviewer: “instead of pod annotations” directly modifies “with a single Kubernetes resource.” It is not a parenthetical aside, so separating it with a comma interrupts the comparison.

```suggestion
The `DatadogInstrumentation` custom resource (CR) lets you configure [Autodiscovery][1] checks and logs with a single Kubernetes resource instead of [pod annotations][2]. With this approach, you can enable, update, and remove integration configurations without editing your Agent or application and triggering a rollout.
```

### Target Services

Targeting a `Service` configures an [endpoint check][6] similar to an annotation on a Kubernetes service.

- Datadog schedules one endpoint check for each endpoint of the Service.
- `%%host%%` resolves to the endpoint IP.
- If an endpoint is backed by a Kubernetes Pod, Datadog adds the Pod tags collected for that Pod.
- If an endpoint is not backed by a Pod, Datadog converts the check into a regular cluster check without Pod-specific tags.

<div class="alert alert-info">

Service targets do not use `containerName`; omit that field.

</div>

Below is an example configuring a nginx check against a Kubernetes `Service`:

```yaml
apiVersion: datadoghq.com/v1alpha1
kind: DatadogInstrumentation
metadata:
  name: <YOUR_CR_NAME>
  namespace: <YOUR_SERVICES_NAMESPACE>
spec:
  targetRef:
    apiVersion: v1
    kind: Service
    name: nginx
  config:
    checks:
      - integration: nginx
        initConfig: {}
        instances:
          - name: "My NGINX Service Endpoints"
            nginx_status_url: "http://%%host%%:%%port%%/status/"
```

## Precedence

> When more than one configuration source applies to a workload, Datadog resolves them in the following order (highest precedence first):

Reviewer: ```suggestion
When more than one configuration source applies to a workload, the Datadog Agent resolves them in the following order (highest precedence first):
```

Use the `DatadogInstrumentation` CR when you want to:

- Configure Autodiscovery checks and logs without modifying workload manifests or adding annotations.
- Use a structured resource spec with validation instead of raw JSON in annotations.
- Centrally manage per-workload Autodiscovery configuration as a dedicated, version-controlled Kubernetes resource.
- Update or remove Autodiscovery configuration without restarting your application pods.

The configuration is reconciled by a controller in the [Datadog Cluster Agent][3]. When you create or update a `DatadogInstrumentation` resource, the Cluster Agent validates the target, reports resource status, and applies the Autodiscovery configuration to the targeted workload.

## Requirements

- Datadog Agent and Cluster Agent v7.82 or later.

To install the CRD and enable the controller, use one of the following:

- Datadog Operator v1.29 or later.

> - Datadog Helm chart v2.223.0 or later.

Author: **TODO:** Update the minimum helm chart version required

Author: https://github.com/DataDog/helm-charts/pull/2836 update pointing to `datadog/datadog` version `3.236.0`

```yaml
apiVersion: datadoghq.com/v1alpha1
kind: DatadogInstrumentation
metadata:
  name: <YOUR_CR_NAME>
  namespace: <YOUR_SERVICES_NAMESPACE>
spec:
  targetRef:
    apiVersion: v1
    kind: Service
    name: nginx
  config:
    checks:
      - integration: nginx
        initConfig: {}
        instances:
          - name: "My NGINX Service Endpoints"
            nginx_status_url: "http://%%host%%:%%port%%/status/"
```

## Precedence

When more than one configuration source applies to a workload, Datadog resolves them in the following order (highest precedence first):

1. Pod annotations
2. `DatadogInstrumentation` resources
3. Static configuration, such as auto-configuration or mounted files

If a workload already has annotation-based Autodiscovery configuration for a check, your `DatadogInstrumentation` configuration does not override it.

## One resource per target

A workload or Service can be the target of only one `DatadogInstrumentation` resource within a namespace. A validation webhook rejects a resource whose `targetRef` already belongs to another resource, or whose `targetRef` points to an unsupported kind.

## Verify scheduled checks

> Resource status confirms that the configuration was accepted. To confirm that the resulting checks are scheduled, run `agent configcheck` in the Node Agent running on a node where the target workload is scheduled.

Reviewer 2: ```suggestion
The resource status shows whether the Cluster Agent accepted the configuration. To verify that the checks are scheduled, run `agent configcheck` on the Node Agent where the target workload runs.
```
